maven-caching/src/main/java/org/apache/maven/caching/CacheControllerImpl.java

```diff
@@ -397,67 +415,126 @@ private RestoredArtifact restoredArtifact( org.apache.maven.artifact.Artifact pa
+                        continue;
+                    }
+
+                    if ( entryName.endsWith( "pom.properties" ) )
+                    {
+                        replaceEntry( jarFile, entry, originalPomPropsVersion, pomPropsVersion, jos );
+                        continue;
+                    }
+
+                    if ( JarFile.MANIFEST_NAME.equals( entryName ) )
+                    {
+                        replaceEntry( jarFile, entry, originalImplVersion, implVersion, jos );
+                        continue;
+                    }
+                    jos.putNextEntry( entry );
+                    try ( InputStream entryInputStream = jarFile.getInputStream( entry ) )
+                    {
+                        int bytesRead;
+                        while ( ( bytesRead = entryInputStream.read( buffer ) ) != -1 )
+                        {
+                            jos.write( buffer, 0, bytesRead );
+                        }
+                    }
+                }
             }
-            LOGGER.debug( "Downloaded artifact " + artifact.getArtifactId() + " to: " + artifactFile );
-            return artifactFile.toFile();
-        } );
-        if ( !cacheConfig.isLazyRestore() )
-        {
-            downloadTask.run();
-            downloadTask.get();
         }
-        return downloadTask;
+        return tmpJarFile.toPath();
     }
 
-    private void putChecksum( Artifact artifact, String projectChecksum )
+    private static void replaceEntry( JarFile jarFile, JarEntry entry,
+                                      String toReplace, String replacement, JarOutputStream jos ) throws IOException
     {
-        final DigestItem projectArtifact = DtoUtils.createdDigestedByProjectChecksum( artifact, projectChecksum );
-        final String dependencyKey = KeyUtils.getArtifactKey( artifact );
-        artifactDigestByKey.put( dependencyKey, projectArtifact );
+        String fullManifest = IOUtils.toString( jarFile.getInputStream( entry ), StandardCharsets.UTF_8.name() );
+        String modified = fullManifest.replaceAll( toReplace, replacement );
 
-        final ArtifactHandler artifactHandler = artifactHandlerManager.getArtifactHandler( artifact.getType() );
-        String packaging = artifactHandler.getPackaging();
-        if ( !Objects.equals( packaging, artifact.getType() ) )
-        {
-            artifactDigestByKey.put( KeyUtils.getArtifactKey( artifact, packaging ), projectArtifact );
-        }
-        String extension = artifactHandler.getExtension();
-        if ( !Objects.equals( extension, artifact.getType() ) && !Objects.equals( extension, packaging ) )
+        byte[] bytes = modified.getBytes( StandardCharsets.UTF_8 );
+        JarEntry newEntry = new JarEntry( entry.getName() );
+        jos.putNextEntry( newEntry );
+        jos.write( bytes );
+    }
+
+    private static boolean isArchive( File file )
```

Comment:
should be izZip and contains("META-INF")

Reply:
> 

what does zip.contains("META-INF") means, should we unpack archive and try to find META-INF inside? Can we have artifacts without META-INF? I suspect yes...

Reply:
ZipFile.stream() doesn't unpack it afair - just gives stream of entries